crates/sp-domains-fraud-proof/src/fraud_proof.rs

```diff
@@ -332,12 +338,60 @@ impl<DomainHash> From<InvalidBundleEquivocationError> for VerificationError<Doma
     }
 }
 
+#[derive(Debug, Encode, Decode, TypeInfo, PartialEq, Eq, Clone)]
+pub enum InvalidBundlesFraudProofType {
+    /// Failed to decode the opaque extrinsic.
+    UndecodableTx(u32),
+    /// Transaction is out of the tx range.
+    OutOfRangeTx(u32),
+    /// Transaction is illegal (unable to pay the fee, etc).
+    IllegalTx(u32, StorageProof),
+    /// Transaction is an invalid XDM
+    InvalidXDM(u32),
+    /// Transaction is an inherent extrinsic.
+    InherentExtrinsic(u32),
+}
+
+impl InvalidBundlesFraudProofType {
+    pub fn invalid_bundle_type(&self) -> InvalidBundleType {
+        match self {
+            InvalidBundlesFraudProofType::UndecodableTx(extrinsic_index) => {
+                InvalidBundleType::UndecodableTx(*extrinsic_index)
+            }
+            InvalidBundlesFraudProofType::OutOfRangeTx(extrinsic_index) => {
+                InvalidBundleType::OutOfRangeTx(*extrinsic_index)
+            }
+            InvalidBundlesFraudProofType::IllegalTx(extrinsic_index, _storage_proof) => {
+                InvalidBundleType::IllegalTx(*extrinsic_index)
+            }
+            InvalidBundlesFraudProofType::InvalidXDM(extrinsic_index) => {
+                InvalidBundleType::InvalidXDM(*extrinsic_index)
+            }
+            InvalidBundlesFraudProofType::InherentExtrinsic(extrinsic_index) => {
+                InvalidBundleType::InherentExtrinsic(*extrinsic_index)
+            }
+        }
+    }
+
+    pub fn extrinsic_index(&self) -> u32 {
+        match self {
+            InvalidBundlesFraudProofType::UndecodableTx(extrinsic_index) => *extrinsic_index,
+            InvalidBundlesFraudProofType::OutOfRangeTx(extrinsic_index) => *extrinsic_index,
+            InvalidBundlesFraudProofType::IllegalTx(extrinsic_index, _storage_proof) => {
+                *extrinsic_index
+            }
+            InvalidBundlesFraudProofType::InvalidXDM(extrinsic_index) => *extrinsic_index,
+            InvalidBundlesFraudProofType::InherentExtrinsic(extrinsic_index) => *extrinsic_index,
+        }
+    }
+}
+
 #[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
 pub struct InvalidBundlesFraudProof<ReceiptHash> {
     pub bad_receipt_hash: ReceiptHash,
     pub domain_id: DomainId,
     pub bundle_index: u32,
-    pub invalid_bundle_type: InvalidBundleType,
+    pub invalid_bundle_fraud_proof_type: InvalidBundlesFraudProofType,
     pub extrinsic_inclusion_proof: StorageProof,
```

Comment:
While `IllegalTx` doesn't need `extrinsic_inclusion_proof` but the execution proof, both of them have the same type of `StorageProof`, instead of introducing `InvalidBundlesFraudProofType` we can reuse this field for `IllegalTx` (surely need better naming). When it is `IllegalTx` fraud proof the verifier use this field as the execution proof otherwise as the `extrinsic_inclusion_proof`.

Reply:
@NingLin-P Hmm. This may require more thought. Given that this is not blocking this PR, is it fine to address this in separate PR?

Reply:
Emmm it is pure refactoring without logical change IMO i.e. not introduce the `invalid_bundle_fraud_proof_type` field and instead reuse the `extrinsic_inclusion_proof` field (with better naming) for the illegal tx

Reply:
@NingLin-P yes. That is what I am saying. The naming part is the hardest part for me. And given that this is pure refactoring without logic change, is it okay to address this in separate PR?

Reply:
@NingLin-P Renamed to `proof_data`, couldnt come up with more generalized name. PTAL.